### ISSUE TITLE
FormattedMessage props

### ISSUE THREAD

**Issue by cknitt** [cca-io/rescript-react-intl#1]
Hi,

First of all, thank you for your work!

It seems you have gone for a different FormattedMessage API than the original react-intl version.

react-intl has

```js
<FormattedMessage id="someId" defaultMessage="the default message" />
```

but your bindings have

```js
<FormattedMessage message=message />
```

with `message` coming from a separate `defineMessages` dictionary.

What's the reasoning behind this change? Personally, I prefer the original API with the message definitions "inline" in the FormattedMessage props.

**Comment by alex35mil** [comment]
@cknitt Hey, prolly our workflow is a bit different. Lemmy explain mine and then you can show yours. Then we can decide on API.

When I need new dictionary I create new locale file and create new dictionary using `defineMessages` function from `react-intl` package: https://github.com/alexfedoseev/bs-react-intl/blob/master/examples/basic/PageLocale.re#L2-L3

`defineMessages` is noop function which's only purpose to help [`babel-plugin-react-intl`](https://github.com/yahoo/babel-plugin-react-intl) extract dictionary items into json translations.

I have [js script](https://gist.github.com/alexfedoseev/8db490d48fd4977cd7e13302c5dcf2d2) that extracts all items into json translations. Of course, this whole thing doesn't work 😂 (yet).

The problem is that `babel-plugin-react-intl` needs this:

```js
import { defineMessages } from "react-intl";
defineMessages({ ... });
```

But BuckleScript generates this:

```js
import * as ReactIntl from "react-intl";
ReactIntl.defineMessages({ ... });
```

So, for now I do manual work putting items in json files.

After these items are added I can use them in UI:

```reason
/* I open locale file so `pageLocale` is available */
open PageLocale;

/* And later on pass message to component */
<ReactIntl.FormattedMessage message=pageLocale##today />
```

I could keep original API but it would require manually passing 2 props:

```reason
open PageLocale;

<ReactIntl.FormattedMessage
  id=pageLocale##today##id
  defaultMessage=pageLocale##today##defaultMessage
/>
```

Which is too verbose and annoying for zero profit (at least for my workflow). In JS they just destructure message object to solve this but I can't do this in Reason so I workarounded it by packing this object into `message` prop.

What's your use case?

**Comment by cknitt** [comment]
@alexfedoseev Thanks for your reply and the information about your workflow.

Our workflow - *in JS* - is as follows:

We use
```js
<FormattedMessage id="someId" defaultMessage="some default message" />
```
where possible and `defineMessages` + `intl.formatMessage()` elsewhere (e.g. for showing an alert from a callback or saga or something like that).

Then we use `babel-plugin-react-intl` plus some scripts to extract the messages. `babel-plugin-react-intl` picks up both the messages defined "inline" in `FormattedMessage` and those from the `defineMessages` calls.

We prefer the "inline" version because that way the declaration and use of a message are kept in one place and it is not possible to forget to remove the declaration when not using it anymore.

I would like to keep this usage for our Reason apps, and I want to extract the messages (both from `FormattedMessage` and `defineMessages`) directly from the Reason sources using a ppx. I have already implemented a proof of concept of such a ppx, but didn't have time to finish it yet.

**Comment by alex35mil** [comment]
@cknitt Seems legit! What about having 2 components for both workflows? `FormattedMessage` will accept `id` & `defaultMessage` props (as in JS). But for second one (with `message` prop) I need a good alt name. Help me to come up w/ something nice and we have a deal? :)

**Comment by alex35mil** [comment]
Also, ppx idea sounds interesting!

**Comment by cknitt** [comment]
Two components to support both workflows would be great! Thanks! 😀

Naming for the second one is hard though. `DefinedMessage`? `FormattedDefinedMessage`? `FormattedMessage2` 😱?  I will let you know if I manage to come up with something better.

Once you have chosen a name, I can also do the PR.

**Comment by cknitt** [comment]
What about `FormattedMessageD` ("D" as in "defined")?

**Comment by alex35mil** [comment]
`DefinedMessage` SGTM 👍

**Comment by cknitt** [comment]
While starting to prepare a PR, I noticed that although you have the latest bs-platform (2.2.2) as a peer and dev dependency, the code is not formatted with the 2.2.2 refmt.

Is that intentional? If not, I would first create a PR that reformats the code with 2.2.2 refmt. (Changes are mostly trailing comma and `(. ...)` instead of `[@bs] (...)` in a few places.)

**Comment by alex35mil** [comment]
Yeah, that's b/c `bs-platform@2.2.2` was released earlier than `reason-cli@3.1.0` and I updated packages due to deprecation warnings from `2.2.2`.

> I would first create a PR that reformats the code with 2.2.2 refmt

That would be awesome!

**Comment by alex35mil** [comment]
`bs-react-intl 0.2.0 published`

Thanks!
